Cleanup moving parts

vdimir opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Bug Fix (user-visible misbehavior in an official stable release)


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Clean moving parts that was merged or mutated during move

#### TODO:

- [x] Try to add stress tests. I managed to reproduce locally but not sure if tests would be stable enough and what to check (probably `system.remote_data_paths`)

vdimir committed:
Cleanup moving parts

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #50489 from ClickHouse/ClickHouse:vdimir/alter_moving_garbage -->
This is an automated comment for commit 4073e6628321688e63da316691413330a60cf2a6 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50489/4073e6628321688e63da316691413330a60cf2a6/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

vdimir commented on src/Storages/MergeTree/MergeTreeData.cpp:
```diff
@@ -7929,15 +7939,15 @@ MovePartsOutcome MergeTreeData::moveParts(const CurrentlyMovingPartsTaggerPtr &
-            if (cloned_part)
-                cloned_part->remove();
+            if (cloned_part.part)
+                cloned_part.part->remove();
```
Probably do not call `remove` explicitly, because `is_temp` flag is set

vdimir committed:
Set temporary_directories_lifetime to integration tests with MOVE

vdimir committed:
upd

vdimir commented on src/Storages/MergeTree/MergeTreePartsMover.cpp:
```diff
@@ -240,14 +243,22 @@ MergeTreeMutableDataPartPtr MergeTreePartsMover::clonePart(const MergeTreeMoveEn
     {
         cloned_part_storage = part->makeCloneOnDisk(disk, MergeTreeData::MOVING_DIR_NAME);
     }
+    String data_part_directory = cloned_part_storage->getFullPath();
+
+    TemporaryClonedPart cloned_part;
+    cloned_part.temporary_directory_lock = data->getTemporaryPartDirectoryHolder(data_part_directory);
```
> Does/should data_part_directory include moving/?

Yes, since we lock part in `moving` directory before it's renamed.

```cpp
cloned_part_storage = part->makeCloneOnDisk(disk, MergeTreeData::MOVING_DIR_NAME);
```

vdimir replied:
Holder should accept only base name of the part https://github.com/ClickHouse/ClickHouse/blob/5d0522cfa6045ccecc77f79dd12b043260e1545b/src/Storages/MergeTree/MergeTreeData.cpp#L2095

Fixed

tavplubix commented on src/Storages/MergeTree/MergeTreePartsMover.cpp:
```diff
@@ -240,14 +243,22 @@ MergeTreeMutableDataPartPtr MergeTreePartsMover::clonePart(const MergeTreeMoveEn
+    String data_part_directory = cloned_part_storage->getFullPath();
+
+    TemporaryClonedPart cloned_part;
+    cloned_part.temporary_directory_lock = data->getTemporaryPartDirectoryHolder(data_part_directory);
```
We should hold the temp dir lock until the part is renamed to persistent name (`renameTo` in `swapClonedPart`)

vdimir committed:
Fixes for cleanup moving parts

vdimir committed:
Add test test_alter_moving_garbage

vdimir commented:
Fixed couple of things:

* One found in flaky check https://s3.amazonaws.com/clickhouse-test-reports/50489/f28c4dd58f600624a07e2c395b4114a8eb8a4272/integration_tests_flaky_check__asan_.html

`getTemporaryPartDirectoryHolder` should accept only basename and also should be taken earlier, before `makeCloneOnDisk` is called.

* Also part was deleted only in `clearOldParts` but not in descrtuctor because of this check

https://github.com/ClickHouse/ClickHouse/blob/5d0522cfa6045ccecc77f79dd12b043260e1545b/src/Storages/MergeTree/IMergeTreeDataPart.cpp#L501-L507

Adjusted check to include parts from `moving/`

* Also added test, locally works fine, let's check in flaky check

vdimir commented on tests/integration/test_alter_moving_garbage/test.py:
```diff
@@ -0,0 +1,90 @@
+            node.query(
+                f"ALTER TABLE test_table MOVE PARTITION '{partition}' TO DISK 's3'",
+            )
+        except QueryRuntimeException as e:
+            # PART_IS_TEMPORARILY_LOCKED
+            assert 384 == e.returncode
```
Is it fine to ignore/retry such an error?

CheSema commented:
Does this solve it? https://github.com/ClickHouse/ClickHouse/issues/49197

vdimir commented:
@CheSema yes, your description is exactly the same

vdimir committed:
Upd test test_alter_moving_garbage

vdimir committed:
Upd test_alter_moving_garbage: use replicated engine

vdimir committed:
Add config param allow_remove_stale_moving_parts

vdimir committed:
Upd test_alter_moving_garbage

vdimir commented:
I added a config parameter (`<allow_remove_stale_moving_parts>`) to enable cleanup only if it's set. Let's merge the PR and wait for a couple of weeks if everything is alright in CI, then we will enable it.
